testutils/clustermanager/gke.go

```diff
@@ -123,55 +145,46 @@ func (gsc *GKESDKClient) setAutoscaling(project, clusterName, location, nodepool
 	return gsc.Service.Projects.Locations.Clusters.NodePools.SetAutoscaling(parent, rb).Do()
 }
 
-// Setup sets up a GKECluster client.
-// minNodes: default to 1 if not provided
-// maxNodes: default to 3 if not provided
-// nodeType: default to n1-standard-4 if not provided
-// region: default to regional cluster if not provided, and use default backup regions
-// zone: default is none, must be provided together with region
-// project: no default
-// addons: cluster addons to be added to cluster
-func (gs *GKEClient) Setup(minNodes *int64, maxNodes *int64, nodeType *string, region *string, zone *string, project *string, addons []string) ClusterOperations {
-	gc := &GKECluster{
-		Request: &GKERequest{
-			MinNodes:      DefaultGKEMinNodes,
-			MaxNodes:      DefaultGKEMaxNodes,
-			NodeType:      DefaultGKENodeType,
-			Region:        DefaultGKERegion,
-			Zone:          DefaultGKEZone,
-			BackupRegions: DefaultGKEBackupRegions,
-			Addons:        addons,
-		},
-	}
+// Setup sets up a GKECluster client, takes GEKRequst as parameter and applies
+// all defaults if not defined.
+func (gs *GKEClient) Setup(r GKERequest) ClusterOperations {
+	gc := &GKECluster{}
 
-	if project != nil { // use provided project and create cluster
-		gc.Project = project
+	if r.Project != "" { // use provided project and create cluster
+		gc.Project = &r.Project
 		gc.NeedCleanup = true
 	}
 
-	if minNodes != nil {
-		gc.Request.MinNodes = *minNodes
-	}
-	if maxNodes != nil {
-		gc.Request.MaxNodes = *maxNodes
+	if r.MinNodes == 0 {
+		r.MinNodes = DefaultGKEMinNodes
 	}
-	if nodeType != nil {
-		gc.Request.NodeType = *nodeType
+	if r.MaxNodes == 0 {
+		r.MaxNodes = DefaultGKEMaxNodes
 	}
-	if region != nil {
-		gc.Request.Region = *region
+	if r.NodeType == "" {
+		r.NodeType = DefaultGKENodeType
 	}
-	if common.GetOSEnv(regionEnv) != "" {
-		gc.Request.Region = common.GetOSEnv(regionEnv)
+	// Only use default backup regions if region is not provided
+	if len(r.BackupRegions) == 0 && r.Region == "" {
```

Comment:
So we won't have `BackupRegions` if we provide `Region`, is it the same logic as the current shell library?

Reply:
It's actually not. Adding this logic as I think it makes more sense, for example in performance test, you do want to keep in the same region so results are comparable, and in this case we shouldn't have hidden `BackupRegions`